File: old/monitor-imessages.js
#!/usr/bin/env node

/**
 * iMessage Monitor - Node.js version
 * Based on Anthropic's iMessage MCP server implementation
 * Monitors for new messages and displays them with rich content support
 */

import { exec } from 'child_process';
import { promisify } from 'util';
import { watch } from 'fs';
import { homedir } from 'os';
import { existsSync, readFileSync, writeFileSync } from 'fs';

const execAsync = promisify(exec);

const DB_PATH = `${homedir()}/Library/Messages/chat.db`;
const STATE_FILE = '/tmp/imessage_monitor_state.json';

// Colors
const colors = {
  reset: '\x1b[0m',
  green: '\x1b[32m',
  blue: '\x1b[34m',
  yellow: '\x1b[33m',
  cyan: '\x1b[36m',
  gray: '\x1b[90m'
};

// Decode attributedBody hex string (from Anthropic MCP server)
function decodeAttributedBody(hexString) {
  try {
    const buffer = Buffer.from(hexString, 'hex');
    const content = buffer.toString();

    const patterns = [
      /NSString">(.*?)</,
      /NSString">([^<]+)/,
      /NSNumber">\d+<.*?NSString">(.*?)</,
      /NSArray">.*?NSString">(.*?)</,
      /"string":\s*"([^"]+)"/,
      /text[^>]*>(.*?)</,
      /message>(.*?)</
    ];

    let text = '';
    for (const pattern of patterns) {
      const match = content.match(pattern);
      if (match?.[1] && match[1].length > 5) {
        text = match[1];
        break;
      }
    }

    const urlMatch = content.match(/(https?:\/\/[^\s<"]+)/);
    const url = urlMatch?.[1];

    if (!text && !url) {
      const readableText = content
        .replace(/streamtyped.*?NSString/g, '')
        .replace(/NSAttributedString.*?NSString/g, '')
        .replace(/NSDictionary.*?$/g, '')
        .replace(/\+[A-Za-z]+\s/g, '')
        .replace(/NSNumber.*?NSValue.*?\*/g, '')
        .replace(/[^\x20-\x7E]/g, ' ')
        .replace(/\s+/g, ' ')
        .trim();

      if (readableText.length > 5) {
        text = readableText;
      } else {
        return { text: '[Rich content]' };
      }
    }

    if (text) {
      text = text
        .replace(/^[+\s]+/, '')
        .replace(/\s*iI\s*[A-Z]\s*$/, '')
        .replace(/\s+/g, ' ')
        .trim();
    }

    return { text: text || url || '', url };
  } catch (error) {
    return { text: '[Message content not readable]' };
  }
}

// Format date
function formatDate(dateStr) {
  const date = new Date(dateStr);
  const now = new Date();
  const isToday = date.toDateString() === now.toDateString();

  if (isToday) {
    return date.toLocaleTimeString('en-US', { hour: '2-digit', minute: '2-digit' });
  } else {
    return date.toLocaleString('en-US', {
      month: 'short',
      day: 'numeric',
      hour: '2-digit',
      minute: '2-digit'
    });
  }
}

// Get state
function getState() {
  if (existsSync(STATE_FILE)) {
    return JSON.parse(readFileSync(STATE_FILE, 'utf8'));
  }
  return { lastMessageId: 0, lastCheck: Date.now() };
}

// Save state
function saveState(state) {
  writeFileSync(STATE_FILE, JSON.stringify(state, null, 2));
}

// Query new messages
async function getNewMessages(lastMessageId) {
  const query = `
    SELECT
      m.ROWID as message_id,
      CASE
        WHEN m.text IS NOT NULL AND m.text != '' THEN m.text
        WHEN m.attributedBody IS NOT NULL THEN hex(m.attributedBody)
        ELSE NULL
      END as content,
      datetime(m.date/1000000000 + strftime('%s', '2001-01-01'), 'unixepoch', 'localtime') as date,
      h.id as sender,
      m.is_from_me,
      m.cache_has_attachments,
      CASE
        WHEN m.text IS NOT NULL AND m.text != '' THEN 0
        WHEN m.attributedBody IS NOT NULL THEN 1
        ELSE 2
      END as content_type
    FROM message m
    INNER JOIN handle h ON h.ROWID = m.handle_id
    WHERE m.ROWID > ${lastMessageId}
      AND (m.text IS NOT NULL OR m.attributedBody IS NOT NULL OR m.cache_has_attachments = 1)
      AND m.item_type = 0
    ORDER BY m.date ASC
  `;

  try {
    const { stdout } = await execAsync(`sqlite3 -json "${DB_PATH}" "${query}"`);

    if (!stdout.trim()) {
      return [];
    }

    const messages = JSON.parse(stdout);

    return messages.map(msg => {
      let content = msg.content || '';

      // Decode attributedBody if needed
      if (msg.content_type === 1) {
        const decoded = decodeAttributedBody(content);
        content = decoded.text;
        if (decoded.url) {
          content += ` [${decoded.url}]`;
        }
      }

      // Check for attachments
      if (msg.cache_has_attachments) {
        content += ' 📎';
      }

      return {
        id: msg.message_id,
        content,
        date: msg.date,
        sender: msg.sender,
        isFromMe: Boolean(msg.is_from_me)
      };
    });
  } catch (error) {
    console.error('Error querying messages:', error.message);
    return [];
  }
}

// Display message
function displayMessage(msg) {
  const time = formatDate(msg.date);
  const sender = msg.isFromMe
    ? `${colors.blue}You${colors.reset}`
    : `${colors.yellow}${msg.sender}${colors.reset}`;

  const arrow = msg.isFromMe ? '→' : '←';

  console.log(`${colors.gray}[${time}]${colors.reset} ${sender} ${arrow} ${msg.content}`);
}

// Send macOS notification
async function sendNotification(count, lastMessage) {
  const title = count === 1 ? 'New iMessage' : `${count} New iMessages`;
  const text = lastMessage.content.substring(0, 100);
  const subtitle = lastMessage.isFromMe ? 'You' : lastMessage.sender;

  const script = `display notification "${text}" with title "${title}" subtitle "${subtitle}"`;

  try {
    await execAsync(`osascript -e '${script}'`);
  } catch (error) {
    // Ignore notification errors
  }
}

// Main monitoring function
async function checkForNewMessages() {
  const state = getState();
  const messages = await getNewMessages(state.lastMessageId);

  if (messages.length > 0) {
    console.log(`\n${colors.green}📨 ${messages.length} new message(s)${colors.reset}\n`);

    messages.forEach(displayMessage);
    console.log('');

    // Update state
    state.lastMessageId = messages[messages.length - 1].id;
    state.lastCheck = Date.now();
    saveState(state);

    // Send notification
    await sendNotification(messages.length, messages[messages.length - 1]);
  }
}

// Initialize
async function initialize() {
  // Check database access
  try {
    await execAsync(`sqlite3 "${DB_PATH}" "SELECT 1;"`);
  } catch (error) {
    console.error('❌ Cannot access Messages database. Grant Full Disk Access to Terminal in System Settings.');
    process.exit(1);
  }

  // Initialize state if needed
  const state = getState();
  if (state.lastMessageId === 0) {
    const { stdout } = await execAsync(`sqlite3 "${DB_PATH}" "SELECT MAX(ROWID) FROM message;"`);
    state.lastMessageId = parseInt(stdout.trim()) || 0;
    saveState(state);
    console.log('✅ Initialized. Monitoring for new messages...\n');
  } else {
    console.log('✅ Monitoring for new messages...\n');
  }
}

// Start monitoring
async function main() {
  await initialize();

  // Watch for database changes
  let debounceTimer;

  watch(DB_PATH, async (eventType) => {
    if (eventType === 'change') {
      // Debounce to avoid multiple rapid checks
      clearTimeout(debounceTimer);
      debounceTimer = setTimeout(async () => {
        await checkForNewMessages();
      }, 500);
    }
  });

  console.log('Press Ctrl+C to stop\n');
}

// Handle exit
process.on('SIGINT', () => {
  console.log('\n\n👋 Stopped monitoring');
  process.exit(0);
});

main().catch(console.error);

File: old/monitor-imessages.py
#!/usr/bin/env python3
"""
iMessage Monitor and Claude Bot
Monitors for self-messages in iMessage and responds using Claude CLI
"""

import sqlite3
import subprocess
import time
import os
import json
from pathlib import Path
from datetime import datetime


class iMessageMonitor:
    def __init__(self):
        self.db_path = Path.home() / "Library/Messages/chat.db"
        self.state_file = Path("/tmp/imessage_claude_state.txt")
        self.poll_interval = 1  # seconds

    def get_last_processed_id(self):
        """Get the ID of the last processed message"""
        if self.state_file.exists():
            return int(self.state_file.read_text().strip())
        # Initialize with current max message ID
        with sqlite3.connect(str(self.db_path)) as conn:
            cursor = conn.execute("SELECT MAX(ROWID) FROM message")
            max_id = cursor.fetchone()[0] or 0
            self.state_file.write_text(str(max_id))
            return max_id

    def update_last_processed_id(self, message_id):
        """Update the last processed message ID"""
        self.state_file.write_text(str(message_id))

    def get_new_self_messages(self, last_id):
        """Query for new messages sent to self"""
        query = """
            SELECT
                m.ROWID as message_id,
                m.text as content,
                datetime(m.date/1000000000 + strftime('%s', '2001-01-01'), 'unixepoch', 'localtime') as date,
                h.id as chat_id,
                m.is_from_me,
                c.chat_identifier
            FROM message m
            LEFT JOIN handle h ON h.ROWID = m.handle_id
            LEFT JOIN chat_message_join cmj ON cmj.message_id = m.ROWID
            LEFT JOIN chat c ON c.ROWID = cmj.chat_id
            WHERE m.ROWID > ?
                AND m.is_from_me = 1
                AND m.text IS NOT NULL
                AND m.text != ''
            ORDER BY m.date ASC
        """

        with sqlite3.connect(str(self.db_path)) as conn:
            conn.row_factory = sqlite3.Row
            cursor = conn.execute(query, (last_id,))
            return [dict(row) for row in cursor.fetchall()]

    def is_self_message(self, message):
        """Check if message is sent to self"""
        # Self messages have chat_identifier that matches user's own number/email
        # Or they appear in a chat with yourself
        return message['is_from_me'] == 1 and message['chat_identifier']

    def execute_claude(self, message_text):
        """Execute claude command and return response"""
        try:
            print(f"🤖 Executing: claude {message_text}")
            result = subprocess.run(
                ['claude', message_text],
                capture_output=True,
                text=True,
                timeout=300  # 5 minute timeout
            )

            if result.returncode == 0:
                return result.stdout.strip()
            else:
                return f"Error: {result.stderr.strip()}"

        except subprocess.TimeoutExpired:
            return "Error: Claude command timed out"
        except Exception as e:
            return f"Error executing Claude: {str(e)}"

    def send_imessage(self, chat_identifier, message):
        """Send an iMessage using AppleScript"""
        # Escape quotes in message
        escaped_message = message.replace('"', '\\"').replace('\\', '\\\\')

        applescript = f'''
        tell application "Messages"
            set targetService to 1st account whose service type = iMessage
            set targetBuddy to participant "{chat_identifier}" of targetService
            send "{escaped_message}" to targetBuddy
        end tell
        '''

        try:
            subprocess.run(
                ['osascript', '-e', applescript],
                check=True,
                capture_output=True,
                text=True
            )
            print(f"✅ Sent reply to {chat_identifier}")
            return True
        except subprocess.CalledProcessError as e:
            print(f"❌ Failed to send message: {e.stderr}")
            return False

    def process_message(self, message):
        """Process a single message"""
        content = message['content']
        chat_id = message['chat_identifier']
        date = message['date']

        print(f"\n📨 [{date}] New self-message: {content}")

        # Execute Claude
        response = self.execute_claude(content)
        print(f"💬 Claude response: {response[:100]}...")

        # Send response back
        if chat_id:
            self.send_imessage(chat_id, response)
        else:
            print("⚠️  No chat identifier found, cannot send reply")

    def monitor(self):
        """Main monitoring loop"""
        print("🔍 iMessage Claude Bot started")
        print("📱 Monitoring for self-messages...")
        print("Press Ctrl+C to stop\n")

        last_id = self.get_last_processed_id()
        print(f"Starting from message ID: {last_id}\n")

        try:
            while True:
                new_messages = self.get_new_self_messages(last_id)

                for message in new_messages:
                    if self.is_self_message(message):
                        self.process_message(message)

                    # Update last processed ID
                    last_id = message['message_id']
                    self.update_last_processed_id(last_id)

                time.sleep(self.poll_interval)

        except KeyboardInterrupt:
            print("\n\n👋 Stopping monitor...")
            print(f"Last processed message ID: {last_id}")


def main():
    monitor = iMessageMonitor()
    monitor.monitor()


if __name__ == "__main__":
    main()
